core/dbt/task/deps.py

```diff
@@ -55,39 +96,149 @@ def track_package_install(
         )
 
     def run(self) -> None:
+        if system.path_exists(self.project.packages_install_path):
+            system.rmtree(self.project.packages_install_path)
+
         system.make_directory(self.project.packages_install_path)
-        packages = self.project.packages.packages
-        if not packages:
+
+        packages_lock_dict = package_data_from_root(self.project.project_root, "package-lock.yml")
+        packages_lock_config = package_config_from_data(packages_lock_dict).packages
+
+        if not packages_lock_config:
             fire_event(DepsNoPackagesFound())
             return
 
         with downloads_directory():
-            final_deps = resolve_packages(packages, self.project, self.cli_vars)
-
+            lock_defined_deps = resolve_lock_packages(packages_lock_config)
             renderer = DbtProjectYamlRenderer(None, self.cli_vars)
 
             packages_to_upgrade = []
-            for package in final_deps:
+
+            for package in lock_defined_deps:
                 package_name = package.name
                 source_type = package.source_type()
                 version = package.get_version()
 
                 fire_event(DepsStartPackageInstall(package_name=package_name))
                 package.install(self.project, renderer)
+
                 fire_event(DepsInstallInfo(version_name=package.nice_version_name()))
+
                 if isinstance(package, RegistryPinnedPackage):
                     version_latest = package.get_version_latest()
+
                     if version_latest != version:
                         packages_to_upgrade.append(package_name)
                         fire_event(DepsUpdateAvailable(version_latest=version_latest))
                     else:
                         fire_event(DepsUpToDate())
+
                 if package.get_subdirectory():
                     fire_event(DepsListSubdirectory(subdirectory=package.get_subdirectory()))
 
                 self.track_package_install(
                     package_name=package_name, source_type=source_type, version=version
                 )
+
             if packages_to_upgrade:
                 fire_event(Formatting(""))
                 fire_event(DepsNotifyUpdatesAvailable(packages=ListOfStrings(packages_to_upgrade)))
+
+
+class LockTask(BaseTask):
+    def __init__(self, args: Any, project: Project):
+        move_to_nearest_project_dir(project.project_root)
+        super().__init__(args=args, config=None, project=project)
+        self.cli_vars = args.vars
+
+    def run(self):
+        lock_filepath = f"{self.project.project_root}/package-lock.yml"
+
+        packages = self.project.packages.packages
+        packages_installed = {"packages": []}
+
+        if not packages:
+            fire_event(DepsNoPackagesFound())
+            return
+
+        with downloads_directory():
+            resolved_deps = resolve_packages(packages, self.project, self.cli_vars)
```

Comment:
Depending on where those packages lives, this might actually pull down a git repo right? Just to understand what's happening, we will pull it down, and remove it in the function above that has comment on it, and pull it down again when install? Did we always pull down git repos twice before?